### ISSUE TITLE
Inquiry Regarding Dataset Availability, Compatibility, and Performance Comparisons

### ISSUE THREAD

**Issue by andrewssobral** [robjinman/richard#2]
Dear Team,

Firstly, I'd like to extend my gratitude for sharing your remarkable work with the community. It's projects like these that inspire and fuel further innovation.

I have a few inquiries that I hope you could assist me with:

1. **Dataset Availability**: Could you provide access to the directory containing the dataset used in your project? Sharing a link or any guidance on how to replicate your results exactly would be immensely helpful.

2. **Platform Compatibility**: Is this project compatible with Mac systems, specifically on both arm64 and Intel architectures? It would be beneficial to know any particularities regarding the setup or execution in these environments.

3. **Performance Metrics**: Do you have any performance or speed comparisons available? Insights into how your solution compares with others, especially in terms of efficiency, would be very valuable.

Thank you in advance for your time and assistance. I look forward to your response.

Warm regards,

Andrews

**Comment by robjinman** [comment]
Hi Andrews, thanks for your interest in richard :)

For most of the testing I've used the following two datasets: [hand-written digits](https://drive.google.com/file/d/1AjwjBiMNnwc4cTfliogVX8hILoKXef9D/view?usp=drive_link), [cat/dog images](https://drive.google.com/file/d/149oPhFCyRl-KEkvm7kIjLidcgOnLbJJe/view?usp=drive_link).

I've not tried building on Mac, but in theory it should be straightforward. With any luck it will just work.

In terms of performance, here I may have to disappoint you. Richard's GPU implementation is currently VERY SLOW when compared to an equivalent keras/tensorflow model. I just tried running the catdog benchmark on my linux desktop, which has a GTX 1070, and, compared to richard running a similar model on the exact same data set, richard was 40X slower than the CUDA-enabled keras version.

However, I've not spent much time optimising richard's GPU implementation at all and there may be some quick wins that will go some way to closing that gap, but I doubt it will ever be competitive with other solutions.

I am curious though how far we could push a vulkan implementation if we got some real GPGPU experts to look at it.

If this project is to serve any practical use it's mainly as an educational resource. The CPU implementation is quite readable and it's easy to find the equations from documentation/math.pdf in the code, whereas most ML frameworks obscure these details.

Rob
